Fix import issue without remote config in the target

sharidas opened the pull request:
Fix the import issue when target does not have
remote config.

- Fixes https://github.com/iterative/dvc/issues/2780

Signed-off-by: Sujith H <sujith.h@gmail.com>

* [x] ❗ Have you followed the guidelines in the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) list?

* [x] 📖 Check this box if this PR **does not** require [documentation](https://dvc.org/doc) updates, or if it does **and** you have created a separate PR in [dvc.org](https://github.com/iterative/dvc.org) with such updates (or at least opened an issue about it in that repo). Please link below to your PR (or issue) in the [dvc.org](https://github.com/iterative/dvc.org) repo.

* [x] ❌ Have you checked DeepSource, CodeClimate, and other sanity checks below? We consider their findings recommendatory and don't expect everything to be addressed. Please review them carefully and fix those that actually improve code or fix bugs.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

sharidas commented:
This PR is still a work in progress. In this PR I have tried to capture the `NoRemoteError` execption and then look for the file/dir inside the dvc repo where we are importing. If the file/dir is found then no further action required as its available in the cache.

I have not added condition to check: what if the cache is not available in the reciever of the import. I need to check the same. 

Any feedback regarding the approach is appreciated.

efiop commented on dvc/dependency/repo.py:
```diff
@@ -71,7 +72,20 @@ def fetch(self):
+                    # else just raise the exception
+                    from dvc.repo import Repo
+
+                    source_repo = Repo()
```
This will create a repo from cwd, which might not be in the repo at all. You are probably looking for `self.repo`.

sharidas replied:
Yah actually I was looking for the same. I have modified the pr to use `self.repo`.

sharidas commented:
Would appreciate review for this PR.

shcheklein commented:
@sharidas should we add a test for this?

mroutis commented on dvc/dependency/repo.py:
```diff
@@ -71,7 +72,15 @@ def fetch(self):
+                try:
+                    repo.cloud.pull(out.get_used_cache())
+                except NoRemoteError:
+                    # Check if the file is available in the cache
```
The comment is describing _what_ the code is doing, but I would appreciate more a comment about _why_ it is doing such thing.

sharidas replied:
I have updated the comment. Let me know if this looks ok?

sharidas commented:
> @sharidas should we add a test for this?

Updated the PR by adding test.

efiop commented on tests/func/test_import.py:
```diff
@@ -56,6 +56,28 @@ def test_import_git_file(erepo_dir, tmp_dir, dvc, scm, src_is_dvc):
+    erepo_dir.scm_gen({src: "hello"}, commit="add a regular file")
+    os.remove(erepo_dir.dvc.config.config_file)
+
+    with tmp_dir.chdir():
```
You are already inside of a tmp_dir, no need to chdir. 🙂

efiop commented on tests/func/test_import.py:
```diff
@@ -56,6 +56,28 @@ def test_import_git_file(erepo_dir, tmp_dir, dvc, scm, src_is_dvc):
+        f = open(dst, "w")
+        f.write("hello")
+        f.close()
+        tmp_dir.dvc.add([fspath(tmp_dir / dst)])
```
You are adding by full path, which will make dvc treat it as an external output. What you want is:
```
tmp_dir.dvc_gen({dst: "hello"})
```
looks like.

efiop replied:
Side note: using open + write + close is a bad style even without pathlib, where you should use `with open()` instead.

efiop commented on tests/func/test_import.py:
```diff
@@ -56,6 +56,28 @@ def test_import_git_file(erepo_dir, tmp_dir, dvc, scm, src_is_dvc):
+        f.write("hello")
+        f.close()
+        tmp_dir.dvc.add([fspath(tmp_dir / dst)])
+        os.remove(tmp_dir / dst)
```
```suggestion
        (tmp_dir / dst).unlink()
```

efiop commented on dvc/dependency/repo.py:
```diff
@@ -72,7 +73,14 @@ def fetch(self):
+                except NoRemoteError:
+                    # It would not be good idea to raise exception if the
+                    # file is already present in the cache
+                    if os.path.exists(self.repo.cache.local.get(out.checksum)):
```
Let's use
```suggestion
                    if not self.repo.cache.local.changed_cache(out.checksum):
```
to make it fully work with directories and be more robust in general. There is nothing wrong with your implementation, just an obscure internal stuff that is more appropriate here.

efiop commented on tests/func/test_import.py:
```diff
@@ -56,6 +56,28 @@ def test_import_git_file(erepo_dir, tmp_dir, dvc, scm, src_is_dvc):
+    dst = "some_file_imported"
+
+    erepo_dir.scm_gen({src: "hello"}, commit="add a regular file")
+    os.remove(erepo_dir.dvc.config.config_file)
```
Hm, I don't think this affects the test at all, because dvc import clones erepo and non-commited changes won't be cloned anyways.

efiop replied:
So what happens in this test currently is that dvc automatically sets default remote to erepo_dir's .dvc/cache and pulls from there. https://github.com/iterative/dvc/blob/master/dvc/external_repo.py#L91 Need to somehow disable it for this test or workaround it. 🤔

sharidas replied:
@efiop If my understanding is correct https://github.com/iterative/dvc/blob/master/tests/dir_helpers.py#L282-L283 is the code which adds remote configuration. Or am I missing something. I didn't quite get the point here. I was thinking: if removal of config from the .dvc/config from the erepo would help to replicate the scenario, while writing the test.

efiop replied:
@sharidasan Oh, good point! I've totally forgotten about that one. We used to use that before https://github.com/iterative/dvc/blob/master/dvc/external_repo.py#L91 logic. But still, as you can see that file is _commited_ and here you are only removing it from the workspace, not committing it, so clone of this erepo would still have that config intact.

sharidas commented:
@efiop I have updated the change again. In the latest change set I have introduced a new fixture ( similar to `erepo_dir` ) which differs from `erepo_dir` where no config data is present. So when the `erepo` is cloned an empty config file is created. Let me know if this looks kind of ok ...

efiop commented on tests/dir_helpers.py:
```diff
@@ -285,3 +293,21 @@ def erepo_dir(tmp_path_factory, monkeypatch):
+
+        path.dvc.close()
+
+    return path
```
Unfortunately removing that `config` part still won't help, as https://github.com/iterative/dvc/blob/master/dvc/external_repo.py#L91 is still in place, which automatically sets it back when cloning 🙁

How about we simply mock `cloud.pull` to raise NoRemoteError in our test?

sharidas replied:
I am bit confused here. When I execute:
```
python -m tests tests/func/test_import.py::test_import_cached_file
```
I do see a config file under location `/tmp/pytest-of-sujith/pytest-0/popen-gw0/erepo0/.dvc/config` in my machine. The file is actually empty. So while cloning it sets back but the content is empty.. Kindly correct me if my obeservation is wrong.

efiop replied:
@sharidasan It creates `.dvc/config.local` though, not `.dvc/config`.

Suor commented on dvc/dependency/repo.py:
```diff
@@ -91,7 +92,14 @@ def fetch(self):
+                except NoRemoteError:
+                    # It would not be good idea to raise exception if the
+                    # file is already present in the cache
+                    if not self.repo.cache.local.changed_cache(out.checksum):
```
I don't think this is a correct logic for dir outs, it should be just `out.changed_cache()`. 

However, I don't understand the overall approach, we just created that repo it has empty cache. What's the point?

Suor replied:
Oh, I see, this is for local import/shared cache case. I would say we should change `.pull()` implementation to check local cache first and only go for upstream when something is missing in local cache.

efiop replied:
@Suor please see https://github.com/iterative/dvc/pull/3120#discussion_r367502238 . out.change_cache() would be better, true.

Suor commented on tests/dir_helpers.py:
```diff
@@ -285,3 +293,21 @@ def erepo_dir(tmp_path_factory, monkeypatch):
+
+
+@pytest.fixture
+def erepo_dir_no_remote_config(tmp_path_factory, monkeypatch):
```
It's an overkill creating a fixture for single test. Any of the following would be better:
- use existing `erepo_dir`, drop upstream in a test
- change `erepo_dir` to not have upstream, fix any issues it causes (if that is not too hard)

efiop replied:
@Suor Def don't need that new fixture, but `change erepo_dir to not have upstream` won't help, see https://github.com/iterative/dvc/pull/3120#discussion_r367500983

sharidas commented:
> I would say we should change .pull() implementation to check local cache first and only go for upstream when something is missing in local cache.

This case is related get/import failure when remote config is missing. Meaning when we get/import the file/dir from the remote url. I am sharing my understanding with respect to the ticket here :)

sharidas commented:
I have updated the test. Also verified the expected code flow to trigger the excecption for `NoRemoteError`. Let me know if this change set looks ok? @efiop @shcheklein
